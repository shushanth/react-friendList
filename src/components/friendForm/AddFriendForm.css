:local(.addForm) {
  display: flex;
  flex: 1;
  flex-direction: column;
}

:local(.addFormGender) {
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 1rem;
  width: 100%;
}
