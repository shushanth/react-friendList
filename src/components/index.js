export { default as AddFriendForm } from "./friendForm/AddFriendForm";
export { default as FriendList } from "./friendList/FriendList";
export {
  default as FriendPagination
} from "./friendPagination/FriendPagination";
