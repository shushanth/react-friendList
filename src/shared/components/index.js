export { default as TextInput } from "./input/TextInput";
export { default as RadioGroup } from "./radio/RadioGroup";
export { default as RadioOption } from "./radio/RadioOption";
export { default as GridRow } from "./grid/GridRow";
export { default as GridColumn } from "./grid/GridColumn";
export { default as Pagination } from "./pagination/Pagination";