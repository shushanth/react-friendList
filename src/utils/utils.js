/**
 * contains utils helpers.
 */

/**
 * class which is used to create enums,
 */
class Enum {
  constructor(obj) {
    const enumHandler = {
      get(target, name) {
        if (target[name]) {
          return target[name];
        }
        throw new Error(`no enum found for ${name}`);
      }
    };
    return new Proxy(Object.freeze(obj), enumHandler);
  }
};

/**
 * keyborad events
 */
export const KEYBOARD_EVENTS = new Enum({
  ENTER_KEY: 13
});

/** TODO : instead of the below, npm package can be used */
export const uuid = () => {
  const s4 = () => {
    return Math.floor((1 + Math.random()) * 0x1000)
      .toString(16)
      .substring(1);
  };
  return (s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4());
};
