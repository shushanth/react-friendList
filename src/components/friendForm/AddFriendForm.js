import React, { Component } from "react";
import classnames from "classnames";
import PropTypes from "prop-types";

import { TextInput, RadioGroup, RadioOption } from "../../shared/components";
import { KEYBOARD_EVENTS } from "../../utils";
import styles from "./AddFriendForm.css";

const { ENTER_KEY } = KEYBOARD_EVENTS;

class AddFriendForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      gender: "Male"
    };
  }

  handleNameChange = event => {
    const { addFriend, pageMove } = this.props;
    let name = event.target.value;
    let gender = this.state.gender;
    this.setState({ name });
    if (name && event.which === ENTER_KEY) {
      addFriend(name, gender);
      this.setState({ name: "", gender: "Male" });
      setTimeout(() => pageMove());
    }
  };

  handleGenderChange = event => {
    let {
      target: { value }
    } = event;
    this.setState(...this.state, { gender: value });
  };

  render() {
    return (
      <div className={classnames(styles.addForm)}>
        <TextInput
          name="addFriend"
          type="text"
          placeholder="Type the name of a friend"
          value={this.state.name}
          focusAuto={true}
          updateOnkeyDown={true}
          onInputChange={this.handleNameChange}
        />
        <div className={classnames(styles.addFormGender)}>
          <RadioGroup
            name="gender"
            radioValue={this.state.gender}
            onRadioChange={this.handleGenderChange}
          >
            <RadioOption value="Male" />
            <RadioOption value="Female" />
          </RadioGroup>
        </div>
      </div>
    );
  }
}

AddFriendForm.PropTypes = {
  addFriend: PropTypes.func
};

export default AddFriendForm;
