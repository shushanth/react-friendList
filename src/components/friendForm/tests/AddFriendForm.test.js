import React from "react";
import { mount, shallow } from "enzyme";

import AddFriendForm from "../AddFriendForm";
import { TextInput, RadioGroup, RadioOption } from "../../../shared/components";
import { KEYBOARD_EVENTS } from "../../../utils";

describe("Component: <AddFriendForm/>", () => {
  const AddFriendFormWapper = shallow(<AddFriendForm />);
  const AddFriendFormMountWrapper = mount(<AddFriendForm />);
  beforeEach(() => {
    const initialState = { name: "", gender: "Male" };
    AddFriendFormMountWrapper.setState(initialState);
  });

  it("should render without crashing", () => {
    expect(AddFriendFormWapper.exists()).toBe(true);
    expect(AddFriendFormWapper).toMatchSnapshot();
  });

  it("should have text input component to add friends", () => {
    expect(AddFriendFormWapper.find("TextInput").length).toBe(1);
  });

  it("should have checkbox components to add gender", () => {
    expect(AddFriendFormWapper.find("RadioGroup").length).toBe(1);
    expect(AddFriendFormWapper.find("RadioOption").length).toBe(2);
  });

  it("should have add friend input with proper props values intially", () => {
    let addFriendInputProps = AddFriendFormWapper.find("TextInput").props();
    expect(addFriendInputProps.value).toEqual("");
    expect(addFriendInputProps.placeholder).toEqual(
      "Type the name of a friend"
    );
    expect(addFriendInputProps.focusAuto).toBeTruthy();
    expect(addFriendInputProps.updateOnkeyDown).toBeTruthy();
  });

  it("should have proper value set for add friend input component, on change of state ", () => {
    AddFriendFormMountWrapper.setState({ name: "Leonardo" });
    expect(AddFriendFormMountWrapper.find("TextInput").props().value).toEqual(
      "Leonardo"
    );
  });

  it("should have checkbox values respective of genders", () => {
    let genderMaleProps = AddFriendFormWapper.find("RadioOption")
      .at(0)
      .props();
    let genderFemaleProps = AddFriendFormWapper.find("RadioOption")
      .at(1)
      .props();
    expect(genderMaleProps.value).toEqual("Male");
    expect(genderFemaleProps.value).toEqual("Female");
  });

  it("should have checkbox selected , with proper state set", () => {
    expect(
      AddFriendFormMountWrapper.find("RadioGroup").props().radioValue
    ).toEqual("Male");
    expect(
      AddFriendFormMountWrapper.find("RadioGroup RadioOption")
        .at(0)
        .props().isChecked
    ).toBeTruthy();
    expect(
      AddFriendFormMountWrapper.find("RadioGroup RadioOption")
        .at(1)
        .props().isChecked
    ).toBeFalsy();
    AddFriendFormMountWrapper.setState({ gender: "Female" });
    expect(
      AddFriendFormMountWrapper.find("RadioGroup").props().radioValue
    ).toEqual("Female");
    expect(
      AddFriendFormMountWrapper.find("RadioGroup RadioOption")
        .at(0)
        .props().isChecked
    ).toBeFalsy();
    expect(
      AddFriendFormMountWrapper.find("RadioGroup RadioOption")
        .at(1)
        .props().isChecked
    ).toBeTruthy();
  });

  describe("method: handleChangeName", () => {
    it("should have proper states updated on invoking this method", () => {
      const AddFriendFormWrapperInstance = AddFriendFormWapper.instance();
      const customEvent = {
        target: { value: "John william" }
      };
      expect(AddFriendFormWapper.state("name")).toEqual("");
      AddFriendFormWrapperInstance.handleNameChange(customEvent);
      expect(AddFriendFormWapper.state("name")).toEqual("John william");
    });
    it("should invoke this method on change of AddFriend input componet", () => {
      const handleChangeNameMock = jest.fn();
      const customEvent = {
        target: { value: "Adam paul" }
      };
      const TextInputComponent = mount(
        <TextInput
          type="text"
          name="addFriend"
          onInputChange={handleChangeNameMock}
        />
      );
      TextInputComponent.find("input[name='addFriend']").simulate(
        "change",
        customEvent
      );
      expect(handleChangeNameMock).toHaveBeenCalled();
    });

    it("should invoke props method addFriend, on enter of text input", () => {
      const addFriendMock = jest.fn();
      const pageMoveMock = jest.fn();
      const customEvent = {
        target: { value: "Adam paul" },
        which: KEYBOARD_EVENTS.ENTER_KEY
      };
      let AddFriendFormComponent = shallow(
        <AddFriendForm addFriend={addFriendMock} pageMove={pageMoveMock} />
      );
      AddFriendFormComponent.instance().handleNameChange(customEvent);
      expect(addFriendMock).toHaveBeenCalledWith("Adam paul", "Male");
    });
  });
  describe("method: handleGenderChange", () => {
    it("should invoke this method and change state should be changed accordingly", () => {
      const handleGenderChangeMock = jest.fn();
      const customEvent = { target: { value: "Female" } };
      const RadioGroupComponent = mount(
        <RadioGroup
          name="gender"
          radioValue="Female"
          onRadioChange={handleGenderChangeMock}
        >
          <RadioOption value="Male" />
          <RadioOption value="Female" />
        </RadioGroup>
      );
      RadioGroupComponent.find("input[type='radio'][value='Male']").simulate(
        "change",
        customEvent
      );
      expect(handleGenderChangeMock).toHaveBeenCalled();
      expect(AddFriendFormWapper.state("gender")).toEqual("Male");
    });
  });
});
