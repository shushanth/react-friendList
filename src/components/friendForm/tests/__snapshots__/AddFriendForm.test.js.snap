exports[`Component: <AddFriendForm/> should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddFriendForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <TextInput
          focusAuto={true}
          name="addFriend"
          onInputChange={[Function]}
          placeholder="Type the name of a friend"
          type="text"
          updateOnkeyDown={true}
          value="" />,
        <div
          className="">
          <RadioGroup
                    name="gender"
                    onRadioChange={[Function]}
                    radioValue="Male">
                    <RadioOption
                              value="Male" />
                    <RadioOption
                              value="Female" />
          </RadioGroup>
</div>,
      ],
      "className": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "focusAuto": true,
          "name": "addFriend",
          "onInputChange": [Function],
          "placeholder": "Type the name of a friend",
          "type": "text",
          "updateOnkeyDown": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <RadioGroup
            name="gender"
            onRadioChange={[Function]}
            radioValue="Male">
            <RadioOption
                        value="Male" />
            <RadioOption
                        value="Female" />
</RadioGroup>,
          "className": "",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <RadioOption
                value="Male" />,
              <RadioOption
                value="Female" />,
            ],
            "name": "gender",
            "onRadioChange": [Function],
            "radioValue": "Male",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "value": "Male",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "value": "Female",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <TextInput
            focusAuto={true}
            name="addFriend"
            onInputChange={[Function]}
            placeholder="Type the name of a friend"
            type="text"
            updateOnkeyDown={true}
            value="" />,
          <div
            className="">
            <RadioGroup
                        name="gender"
                        onRadioChange={[Function]}
                        radioValue="Male">
                        <RadioOption
                                    value="Male" />
                        <RadioOption
                                    value="Female" />
            </RadioGroup>
</div>,
        ],
        "className": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "focusAuto": true,
            "name": "addFriend",
            "onInputChange": [Function],
            "placeholder": "Type the name of a friend",
            "type": "text",
            "updateOnkeyDown": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <RadioGroup
              name="gender"
              onRadioChange={[Function]}
              radioValue="Male">
              <RadioOption
                            value="Male" />
              <RadioOption
                            value="Female" />
</RadioGroup>,
            "className": "",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <RadioOption
                  value="Male" />,
                <RadioOption
                  value="Female" />,
              ],
              "name": "gender",
              "onRadioChange": [Function],
              "radioValue": "Male",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "value": "Male",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "value": "Female",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
